src/Features/Protocol/LanguageServices/RoslynLanguageService.cs

```diff
@@ -0,0 +1,673 @@
+                if (location == null)
+                {
+                    return null;
+                }
+
+                var symbol = await GetSymbolAsync(location).ConfigureAwait(false);
+                if (symbol == null)
+                {
+                    return null;
+                }
+
+                return new DocumentSymbol
+                {
+                    Name = symbol.Name,
+                    Detail = item.Text,
+                    Kind = item.Glyph.ToSymbolKind(),
+                    Deprecated = symbol.GetAttributes().Any(x => x.AttributeClass.MetadataName == "ObsoleteAttribute"),
+                    Range = item.Spans.First().ToRange(text),
+                    SelectionRange = location.SourceSpan.ToRange(text),
+                    Children = await GetChildrenAsync(item.ChildItems).ConfigureAwait(false),
+                };
+            }
+
+            Location GetLocation(NavigationBarItem item)
+            {
+                if (!(item is NavigationBarSymbolItem symbolItem))
+                {
+                    return null;
+                }
+
+                var symbols = symbolItem.NavigationSymbolId.Resolve(compilation, cancellationToken: cancellationToken);
+                var symbol = symbols.Symbol;
+
+                if (symbol == null)
+                {
+                    if (symbolItem.NavigationSymbolIndex < symbols.CandidateSymbols.Length)
+                    {
+                        symbol = symbols.CandidateSymbols[symbolItem.NavigationSymbolIndex.Value];
+                    }
+                    else
+                    {
+                        return null;
+                    }
+                }
+
+                var location = symbol.Locations.FirstOrDefault(l => l.SourceTree.Equals(tree));
+                return location ?? symbol.Locations.FirstOrDefault();
+            }
+
+            async Task<List<DocumentSymbol>> GetChildrenAsync(IEnumerable<NavigationBarItem> items)
+            {
+                var list = new List<DocumentSymbol>();
+                foreach (var item in items)
+                {
+                    list.Add(await CreateDocumentSymbolAsync(item).ConfigureAwait(false));
+                }
+
+                return list;
+            }
+
+            async Task<ISymbol> GetSymbolAsync(Location location)
+            {
+                var model = compilation.GetSemanticModel(location.SourceTree);
+                var root = await model.SyntaxTree.GetRootAsync(cancellationToken).ConfigureAwait(false);
+                var node = root.FindNode(location.SourceSpan);
+
+                while (node != null)
+                {
+                    var symbol = model.GetDeclaredSymbol(node);
+                    if (symbol != null)
+                    {
+                        return symbol;
+                    }
+
+                    node = node.Parent;
+                }
+
+                return null;
+            }
+        }
```

Comment:
a massive funciton with tons of local functions is just really hard to read and grok.  can you just extract a normal class?  htis would fit far more with how the roslyn ide generally does all of this sort of thing.